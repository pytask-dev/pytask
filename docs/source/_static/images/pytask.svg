<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 24.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 438.9 438.9" style="enable-background:new 0 0 438.9 438.9;" xml:space="preserve">
<style type="text/css">
	.st0{fill:#C19C00;}
	.st1{fill:#13A10E;}
	.st2{fill:#C50F1F;}
</style>
<path d="M348,57.5h-39.7V39.7c0-5.7-6.3-8.4-12-8.4h-30.8C258.1,10.5,239.8,0,218.9,0c-20.7-0.8-39.5,11.9-46.5,31.3h-30.3
	c-5.7,0-11.5,2.6-11.5,8.4v17.8H90.9C67.4,57.8,48.1,76.3,47,99.8v299.4c0,23,20.9,39.7,43.9,39.7h257c23,0,43.9-16.7,43.9-39.7
	V99.8C390.7,76.3,371.5,57.8,348,57.5z M151.5,52.3h28.7c5-0.6,9-4.4,9.9-9.4c3.1-13.5,14.9-23.1,28.7-23.5
	c13.7,0.4,25.3,10.1,28.2,23.5c0.9,5.1,5.2,9,10.4,9.4h29.8v41.8H151.5V52.3z M371,399.2c0,11.5-11.5,18.8-23,18.8h-257
	c-11.5,0-23-7.3-23-18.8V99.8c1.1-12,11-21.2,23-21.4h39.7V105c0.6,5.9,5.6,10.2,11.5,9.9h154.1c6,0.3,11.2-4,12-9.9V78.4H348
	c12,0.2,21.9,9.5,23,21.4L371,399.2L371,399.2z"/>
<path class="st0" d="M179.2,233.6c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,250.3c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,234,179.4,233.8,179.2,233.6z"/>
<path d="M329.2,256H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,256,329.2,256z"/>
<path class="st1" d="M179.2,150c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,166.7c-3.9-4.1-10.4-4.4-14.6-0.5c-4,4.2-4,10.9,0,15.2
	l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14C179.6,150.4,179.4,150.2,179.2,150z"/>
<path d="M329.2,172.4H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,172.4,329.2,172.4z"
	/>
<path class="st2" d="M179.2,317.2c-3.9-4.1-10.4-4.4-14.6-0.5l-33.4,31.9L117,333.9c-3.9-4.1-10.4-4.4-14.6-0.5
	c-4,4.2-4,10.9,0,15.2l21.4,21.9c1.9,2.1,4.5,3.2,7.3,3.1c2.8,0,5.4-1.2,7.3-3.1l40.8-38.7c4-3.7,4.3-10,0.6-14
	C179.6,317.6,179.4,317.4,179.2,317.2z"/>
<path d="M329.2,339.6H209c-5.8,0-10.4,4.7-10.4,10.4s4.7,10.4,10.4,10.4h120.2c5.8,0,10.4-4.7,10.4-10.4S334.9,339.6,329.2,339.6z"
	/>
</svg>
