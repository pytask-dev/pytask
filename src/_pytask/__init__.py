"""This module should not contain any imports except for the version."""
from __future__ import annotations


try:
    from ._version import version as __version__
except ImportError:  # pragma: no cover
    # broken installation, we don't even try unknown only works because we do poor mans
    # version compare
    __version__ = "unknown"


__all__ = ["__version__"]
