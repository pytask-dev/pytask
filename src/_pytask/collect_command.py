"""This module contains the implementation of ``pytask collect``."""
from __future__ import annotations

import sys
from pathlib import Path
from typing import Any
from typing import TYPE_CHECKING

import click
from _pytask.click import ColoredCommand
from _pytask.config import hookimpl
from _pytask.console import console
from _pytask.console import create_url_style_for_path
from _pytask.console import FILE_ICON
from _pytask.console import format_task_id
from _pytask.console import PYTHON_ICON
from _pytask.console import TASK_ICON
from _pytask.exceptions import CollectionError
from _pytask.exceptions import ConfigurationError
from _pytask.exceptions import ResolvingDependenciesError
from _pytask.mark import select_by_keyword
from _pytask.mark import select_by_mark
from _pytask.outcomes import ExitCode
from _pytask.path import find_common_ancestor
from _pytask.path import relative_to
from _pytask.pluginmanager import get_plugin_manager
from _pytask.session import Session
from pybaum.tree_util import tree_just_flatten
from rich.text import Text
from rich.tree import Tree


if TYPE_CHECKING:
    from typing import NoReturn
    from _pytask.nodes import Task


@hookimpl(tryfirst=True)
def pytask_extend_command_line_interface(cli: click.Group) -> None:
    """Extend the command line interface."""
    cli.add_command(collect)


@hookimpl
def pytask_parse_config(
    config: dict[str, Any], config_from_cli: dict[str, Any]
) -> None:
    """Parse configuration."""
    config["nodes"] = config_from_cli.get("nodes", False)


@click.command(cls=ColoredCommand)
@click.option("--nodes", is_flag=True, help="Show a task's dependencies and products.")
def collect(**config_from_cli: Any | None) -> NoReturn:
    """Collect tasks and report information about them."""
    config_from_cli["command"] = "collect"

    try:
        # Duplication of the same mechanism in :func:`pytask.main.main`.
        pm = get_plugin_manager()
        from _pytask import cli

        pm.register(cli)
        pm.hook.pytask_add_hooks(pm=pm)

        config = pm.hook.pytask_configure(pm=pm, config_from_cli=config_from_cli)
        session = Session.from_config(config)

    except (ConfigurationError, Exception):
        session = Session({}, None)
        session.exit_code = ExitCode.CONFIGURATION_FAILED
        console.print_exception()

    else:
        try:
            session.hook.pytask_log_session_header(session=session)
            session.hook.pytask_collect(session=session)
            session.hook.pytask_resolve_dependencies(session=session)

            tasks = _select_tasks_by_expressions_and_marker(session)

            common_ancestor = _find_common_ancestor_of_all_nodes(
                tasks, session.config["paths"], session.config["nodes"]
            )
            dictionary = _organize_tasks(tasks)
            if dictionary:
                _print_collected_tasks(
                    dictionary,
                    session.config["nodes"],
                    session.config["editor_url_scheme"],
                    common_ancestor,
                )

            console.print()
            console.rule(style="neutral")

        except CollectionError:
            session.exit_code = ExitCode.COLLECTION_FAILED

        except ResolvingDependenciesError:
            session.exit_code = ExitCode.RESOLVING_DEPENDENCIES_FAILED

        except Exception:
            session.exit_code = ExitCode.FAILED
            console.print_exception()
            console.rule(style="failed")

    sys.exit(session.exit_code)


def _select_tasks_by_expressions_and_marker(session: Session) -> list[Task]:
    """Select tasks by expressions and marker."""
    all_tasks = {task.name for task in session.tasks}
    remaining_by_mark = select_by_mark(session, session.dag) or all_tasks
    remaining_by_keyword = select_by_keyword(session, session.dag) or all_tasks
    remaining = remaining_by_mark & remaining_by_keyword

    return [task for task in session.tasks if task.name in remaining]


def _find_common_ancestor_of_all_nodes(
    tasks: list[Task], paths: list[Path], show_nodes: bool
) -> Path:
    """Find common ancestor from all nodes and passed paths."""
    all_paths = []
    for task in tasks:
        all_paths.append(task.path)
        if show_nodes:
            all_paths.extend(x.path for x in tree_just_flatten(task.depends_on))
            all_paths.extend(x.path for x in tree_just_flatten(task.produces))

    common_ancestor = find_common_ancestor(*all_paths, *paths)

    return common_ancestor


def _organize_tasks(tasks: list[Task]) -> dict[Path, list[Task]]:
    """Organize tasks in a dictionary.

    The dictionary has file names as keys and then a dictionary with task names and
    below a dictionary with dependencies and targets.

    """
    dictionary: dict[Path, list[Task]] = {}
    for task in tasks:
        dictionary[task.path] = dictionary.get(task.path, [])
        dictionary[task.path].append(task)

    sorted_dict = {}
    for k in sorted(dictionary):
        sorted_dict[k] = sorted(dictionary[k], key=lambda x: x.path)

    return sorted_dict


def _print_collected_tasks(
    dictionary: dict[Path, list[Task]],
    show_nodes: bool,
    editor_url_scheme: str,
    common_ancestor: Path,
) -> None:
    """Print the information on collected tasks.

    Parameters
    ----------
    dictionary : Dict[Path, List["Task"]]
        A dictionary with path on the first level, tasks on the second, dependencies and
        products on the third.
    show_nodes : bool
        Indicator for whether dependencies and products should be displayed.
    editor_url_scheme : str
        The scheme to create an url.
    common_ancestor : Path
        The path common to all tasks and nodes.

    """
    # Have a new line between the number of collected tasks and this info.
    console.print()

    tree = Tree("Collected tasks:", highlight=True)

    for module, tasks in dictionary.items():
        reduced_module = relative_to(module, common_ancestor)
        url_style = create_url_style_for_path(module, editor_url_scheme)
        module_branch = tree.add(
            Text.assemble(
                PYTHON_ICON, "<Module ", Text(str(reduced_module), style=url_style), ">"
            )
        )

        for task in tasks:
            reduced_task_name = format_task_id(
                task,
                editor_url_scheme=editor_url_scheme,
                short_name=True,
                relative_to=common_ancestor,
            )
            task_branch = module_branch.add(
                Text.assemble(TASK_ICON, "<Function ", reduced_task_name, ">"),
            )

            if show_nodes:
                for node in sorted(
                    tree_just_flatten(task.depends_on), key=lambda x: x.path
                ):
                    reduced_node_name = relative_to(node.path, common_ancestor)
                    url_style = create_url_style_for_path(node.path, editor_url_scheme)
                    task_branch.add(
                        Text.assemble(
                            FILE_ICON,
                            "<Dependency ",
                            Text(str(reduced_node_name), style=url_style),
                            ">",
                        )
                    )

                for node in sorted(
                    tree_just_flatten(task.produces), key=lambda x: x.path
                ):
                    reduced_node_name = relative_to(node.path, common_ancestor)
                    url_style = create_url_style_for_path(node.path, editor_url_scheme)
                    task_branch.add(
                        Text.assemble(
                            FILE_ICON,
                            "<Product ",
                            Text(str(reduced_node_name), style=url_style),
                            ">",
                        )
                    )

    console.print(tree)
