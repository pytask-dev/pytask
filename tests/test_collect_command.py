from __future__ import annotations

import os
import pickle
import textwrap
from pathlib import Path

import pytest
from _pytask.collect_command import _find_common_ancestor_of_all_nodes
from _pytask.collect_command import _print_collected_tasks
from _pytask.nodes import PathNode
from attrs import define
from pytask import cli
from pytask import ExitCode
from pytask import Task


@pytest.mark.end_to_end()
def test_collect_task(runner, tmp_path):
    source = """
    import pytask

    @pytask.mark.depends_on("in.txt")
    @pytask.mark.produces("out.txt")
    def task_example():
        pass
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in.txt").touch()

    result = runner.invoke(cli, ["collect", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured

    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "--nodes"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured
    assert "<Dependency" in captured
    assert "in.txt>" in captured
    assert "<Product" in captured
    assert "out.txt>" in captured


@pytest.mark.end_to_end()
def test_collect_task_new_interface(runner, tmp_path):
    source = """
    import pytask
    from pathlib import Path

    def task_example(depends_on=Path("in.txt"), arg=1, produces=Path("out.txt")):
        pass
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in.txt").touch()

    result = runner.invoke(cli, ["collect", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured

    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "--nodes"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured
    assert "<Dependency" in captured
    assert "in.txt>" in captured
    assert "<Product" in captured
    assert "out.txt>" in captured
    assert "arg" in captured


@pytest.mark.end_to_end()
def test_collect_task_in_root_dir(runner, tmp_path):
    source = """
    import pytask

    @pytask.mark.depends_on("in.txt")
    @pytask.mark.produces("out.txt")
    def task_example():
        pass
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in.txt").touch()

    cwd = Path.cwd()
    os.chdir(tmp_path)
    result = runner.invoke(cli, ["collect"])
    os.chdir(cwd)

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured


@pytest.mark.end_to_end()
def test_collect_parametrized_tasks(runner, tmp_path):
    source = """
    import pytask

    for arg, produces in [(0, "out_0.txt"), (1, "out_1.txt")]:

        @pytask.mark.task
        @pytask.mark.depends_on("in.txt")
        def task_example(arg=arg, produces=produces):
            pass
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in.txt").touch()

    result = runner.invoke(cli, ["collect", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "").replace("\u2502", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "[0-out_0.txt]>" in captured
    assert "[1-out_1.txt]>" in captured


@pytest.mark.end_to_end()
def test_collect_task_with_expressions(runner, tmp_path):
    source = """
    import pytask

    @pytask.mark.depends_on("in_1.txt")
    @pytask.mark.produces("out_1.txt")
    def task_example_1():
        pass

    @pytask.mark.depends_on("in_2.txt")
    @pytask.mark.produces("out_2.txt")
    def task_example_2():
        pass
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in_1.txt").touch()
    tmp_path.joinpath("in_2.txt").touch()

    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "-k", "_1"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Function" in captured
    assert "task_example_2>" not in captured

    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "-k", "_1", "--nodes"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Dependency" in captured
    assert "in_1.txt>" in captured
    assert "<Product" in captured
    assert "out_1.txt>" in captured


@pytest.mark.end_to_end()
def test_collect_task_with_marker(runner, tmp_path):
    source = """
    import pytask

    @pytask.mark.wip
    @pytask.mark.depends_on("in_1.txt")
    @pytask.mark.produces("out_1.txt")
    def task_example_1():
        pass

    @pytask.mark.depends_on("in_2.txt")
    @pytask.mark.produces("out_2.txt")
    def task_example_2():
        pass
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in_1.txt").touch()

    config = """
    [tool.pytask.ini_options]
    markers = {'wip' = 'A work-in-progress marker.'}
    """
    tmp_path.joinpath("pyproject.toml").write_text(textwrap.dedent(config))

    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "-m", "wip"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Function" in captured
    assert "task_example_2>" not in captured

    result = runner.invoke(
        cli, ["collect", tmp_path.as_posix(), "-m", "wip", "--nodes"]
    )

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Dependency" in captured
    assert "in_1.txt>" in captured
    assert "<Product" in captured
    assert "out_1.txt>" in captured


@pytest.mark.end_to_end()
def test_collect_task_with_ignore_from_config(runner, tmp_path):
    source = """
    import pytask

    @pytask.mark.depends_on("in_1.txt")
    @pytask.mark.produces("out_1.txt")
    def task_example_1():
        pass
    """
    tmp_path.joinpath("task_example_1.py").write_text(textwrap.dedent(source))

    source = """
    @pytask.mark.depends_on("in_2.txt")
    @pytask.mark.produces("out_2.txt")
    def task_example_2():
        pass
    """
    tmp_path.joinpath("task_example_2.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in_1.txt").touch()

    config = """
    [tool.pytask.ini_options]
    ignore = ["task_example_2.py"]
    """
    tmp_path.joinpath("pyproject.toml").write_text(textwrap.dedent(config))

    result = runner.invoke(cli, ["collect", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_example_1.py>" in captured
    assert "task_example_2.py>" not in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Function" in captured
    assert "task_example_2>" not in captured

    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "--nodes"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_example_1.py>" in captured
    assert "task_example_2.py>" not in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Dependency" in captured
    assert "in_1.txt>" in captured
    assert "<Product" in captured
    assert "out_1.txt>" in captured


@pytest.mark.end_to_end()
def test_collect_task_with_ignore_from_cli(runner, tmp_path):
    source = """
    import pytask

    @pytask.mark.depends_on("in_1.txt")
    @pytask.mark.produces("out_1.txt")
    def task_example_1():
        pass
    """
    tmp_path.joinpath("task_example_1.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in_1.txt").touch()

    source = """
    @pytask.mark.depends_on("in_2.txt")
    @pytask.mark.produces("out_2.txt")
    def task_example_2():
        pass
    """
    tmp_path.joinpath("task_example_2.py").write_text(textwrap.dedent(source))

    result = runner.invoke(
        cli, ["collect", tmp_path.as_posix(), "--ignore", "task_example_2.py"]
    )

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_example_1.py>" in captured
    assert "task_example_2.py>" not in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Function" in captured
    assert "task_example_2>" not in captured

    result = runner.invoke(
        cli,
        ["collect", tmp_path.as_posix(), "--ignore", "task_example_2.py", "--nodes"],
    )

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_example_1.py>" in captured
    assert "task_example_2.py>" not in captured
    assert "<Function" in captured
    assert "task_example_1>" in captured
    assert "<Dependency" in captured
    assert "in_1.txt>" in captured
    assert "<Product" in captured
    assert "out_1.txt>" in captured


@define
class Node:
    path: Path

    def state(self):
        ...


def function(depends_on, produces):  # noqa: ARG001
    ...


@pytest.mark.unit()
def test_print_collected_tasks_without_nodes(capsys):
    dictionary = {
        Path("task_path.py"): [
            Task(
                base_name="function",
                path=Path("task_path.py"),
                function=function,
                depends_on={0: Node("in.txt")},
                produces={0: Node("out.txt")},
            )
        ]
    }

    _print_collected_tasks(dictionary, False, "file", Path())

    captured = capsys.readouterr().out
    assert "<Module task_path.py>" in captured
    assert "<Function task_path.py::function>" in captured
    assert "<Dependency in.txt>" not in captured
    assert "<Product out.txt>" not in captured


@pytest.mark.unit()
def test_print_collected_tasks_with_nodes(capsys):
    dictionary = {
        Path("task_path.py"): [
            Task(
                base_name="function",
                path=Path("task_path.py"),
                function=function,
                depends_on={"depends_on": PathNode(name="in.txt", path=Path("in.txt"))},
                produces={0: PathNode(name="out.txt", path=Path("out.txt"))},
            )
        ]
    }

    _print_collected_tasks(dictionary, True, "file", Path())

    captured = capsys.readouterr().out

    assert "<Module task_path.py>" in captured
    assert "<Function task_path.py::function>" in captured
    assert "<Dependency in.txt>" in captured
    assert "<Product out.txt>" in captured


@pytest.mark.unit()
@pytest.mark.parametrize(("show_nodes", "expected_add"), [(False, "src"), (True, "..")])
def test_find_common_ancestor_of_all_nodes(show_nodes, expected_add):
    tasks = [
        Task(
            base_name="function",
            path=Path.cwd() / "src" / "task_path.py",
            function=function,
            depends_on={
                "depends_on": PathNode.from_path(Path.cwd() / "src" / "in.txt")
            },
            produces={
                0: PathNode.from_path(
                    Path.cwd().joinpath("..", "bld", "out.txt").resolve()
                )
            },
        )
    ]

    result = _find_common_ancestor_of_all_nodes(tasks, [Path.cwd() / "src"], show_nodes)
    assert result == Path.cwd().joinpath(expected_add).resolve()


@pytest.mark.end_to_end()
def test_task_name_is_shortened(runner, tmp_path):
    tmp_path.joinpath("a", "b").mkdir(parents=True)
    tmp_path.joinpath("a", "b", "task_example.py").write_text("def task_example(): ...")

    result = runner.invoke(cli, ["collect", tmp_path.as_posix()])

    assert result.exit_code == ExitCode.OK
    assert "task_example.py::task_example" in result.output
    assert "a/b/task_example.py::task_example" not in result.output


@pytest.mark.end_to_end()
def test_python_node_is_collected(runner, tmp_path):
    source = """
    from pytask import Product
    from typing_extensions import Annotated
    from pathlib import Path

    def task_example(
        dependency: int = 1, path: Annotated[Path, Product] = Path("out.txt")
    ):
        ...
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "--nodes"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured
    assert "Dependency" in captured
    assert "Product" in captured


@pytest.mark.end_to_end()
def test_none_is_a_python_node(runner, tmp_path):
    source = """
    from pytask import Product
    from typing_extensions import Annotated
    from pathlib import Path

    def task_example(
        dependency = None, path: Annotated[Path, Product] = Path("out.txt")
    ):
        ...
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "--nodes"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured
    assert "Dependency" in result.output
    assert "Product" in captured


@pytest.mark.end_to_end()
def test_python_nodes_are_aggregated_into_one(runner, tmp_path):
    source = """
    from pytask import Product
    from typing_extensions import Annotated
    from pathlib import Path

    def task_example(
        nested = {"a": (1, 2), 2: {"b": None}},
        path: Annotated[Path, Product] = Path("out.txt")
    ):
        ...
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    result = runner.invoke(cli, ["collect", tmp_path.as_posix(), "--nodes"])

    assert result.exit_code == ExitCode.OK
    captured = result.output.replace("\n", "").replace(" ", "")
    assert "<Module" in captured
    assert "task_module.py>" in captured
    assert "<Function" in captured
    assert "task_example>" in captured
    assert "Dependency" in result.output
    assert "Product" in captured


@pytest.mark.end_to_end()
def test_node_protocol_for_custom_nodes(runner, tmp_path):
    source = """
    from typing_extensions import Annotated
    from pytask import Product
    from attrs import define
    from pathlib import Path

    @define
    class CustomNode:
        name: str
        value: str

        def state(self):
            return self.value

        def load(self): ...
        def save(self, value): ...

    def task_example(
        data = CustomNode("custom", "text"),
        out: Annotated[Path, Product] = Path("out.txt"),
    ) -> None:
        out.write_text(data)
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, ["collect", "--nodes", tmp_path.as_posix()])
    assert result.exit_code == ExitCode.OK
    assert "<Dependency custom>" in result.output


@pytest.mark.end_to_end()
def test_node_protocol_for_custom_nodes_with_paths(runner, tmp_path):
    source = """
    from typing_extensions import Annotated
    from pytask import Product
    from pathlib import Path
    from attrs import define
    import pickle

    @define
    class PickleFile:
        name: str
        path: Path

        def state(self):
            return str(self.path.stat().st_mtime)

        def load(self):
            with self.path.open("rb") as f:
                out = pickle.load(f)
            return out

        def save(self, value):
            with self.path.open("wb") as f:
                pickle.dump(value, f)

    _PATH = Path(__file__).parent.joinpath("in.pkl")

    def task_example(
        data = PickleFile(_PATH.as_posix(), _PATH),
        out: Annotated[Path, Product] = Path("out.txt"),
    ) -> None:
        out.write_text(data)
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    tmp_path.joinpath("in.pkl").write_bytes(pickle.dumps("text"))

    result = runner.invoke(cli, ["collect", "--nodes", tmp_path.as_posix()])
    assert result.exit_code == ExitCode.OK
    assert "in.pkl" in result.output


@pytest.mark.end_to_end()
def test_setting_name_for_python_node_via_annotation(runner, tmp_path):
    source = """
    from pathlib import Path
    from typing_extensions import Annotated
    from pytask import Product, PythonNode
    from typing import Any

    def task_example(
        input: Annotated[str, PythonNode(name="node-name", value="text")],
        path: Annotated[Path, Product] = Path("out.txt"),
    ) -> None:
        path.write_text(input)
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))

    result = runner.invoke(cli, ["collect", "--nodes", tmp_path.as_posix()])
    assert result.exit_code == ExitCode.OK
    assert "Dependency" in result.output


@pytest.mark.end_to_end()
def test_more_nested_pytree_and_python_node_as_return(runner, tmp_path):
    source = """
    from pathlib import Path
    from typing import Any
    from typing_extensions import Annotated
    from pytask import PythonNode
    from typing import Dict

    nodes = [
        PythonNode(name="dict"),
        (PythonNode(name="tuple1"), PythonNode(name="tuple2")),
        PythonNode(name="int")
    ]

    def task_example() -> Annotated[Dict[str, str], nodes]:
        return [{"first": "a", "second": "b"}, (1, 2), 1]
    """
    tmp_path.joinpath("task_module.py").write_text(textwrap.dedent(source))
    result = runner.invoke(cli, ["collect", "--nodes", tmp_path.as_posix()])
    assert result.exit_code == ExitCode.OK
    assert "return::0" in result.output
    assert "return::1-0" in result.output
    assert "return::1-1" in result.output
    assert "return::2" in result.output
