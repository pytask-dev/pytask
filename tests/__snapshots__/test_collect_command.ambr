# serializer version: 1
# name: test_more_nested_pytree_and_python_node_as_return
  '''
  ───────────────────────────── Start pytask session ─────────────────────────────
  Platform: <platform> -- Python <version>, pytask <version>, pluggy <version>
  Root: <path>
  Collected 1 task.
  
  Collected tasks:
  └── 🐍 <Module test_more_nested_pytree_and_py0/task_module.py>
      └── 📝 <Function task_module.py::task_example>
          ├── 📄 <Product 
          │   test_more_nested_pytree_and_py0/task_module.py::task_example::return
          │   ::0>
          ├── 📄 <Product 
          │   test_more_nested_pytree_and_py0/task_module.py::task_example::return
          │   ::1-0>
          ├── 📄 <Product 
          │   test_more_nested_pytree_and_py0/task_module.py::task_example::return
          │   ::1-1>
          └── 📄 <Product 
              test_more_nested_pytree_and_py0/task_module.py::task_example::return
              ::2>
  
  ────────────────────────────────────────────────────────────────────────────────
  '''
# ---
